packages/roosterjs-editor-dom/lib/utils/adjustUserPosition.ts

```diff
@@ -0,0 +1,251 @@
+import { NodeType, PositionType, QueryScope } from 'roosterjs-editor-types';
+import { safeInstanceOf } from 'roosterjs';
+import {
+    changeElementTag,
+    contains,
+    createRange,
+    findClosestElementAncestor,
+    getBlockElementAtNode,
+    getTagOfNode,
+    isNodeEmpty,
+    isPositionAtBeginningOf,
+    isVoidHtmlElement,
+    LinkInlineElement,
+    Position,
+    PositionContentSearcher,
+    queryElements,
+    splitBalancedNodeRange,
+    splitTextNode,
+    toArray,
+    unwrap,
+    VTable,
+    wrap,
+} from 'roosterjs-editor-dom';
+
+/**
+ * Adjust position for A tag don't be nested inside another A tag.
+ */
+function adjustInsertPositionForHyperLink(
+    root: HTMLElement,
+    nodeToInsert: Node,
+    position: Position,
+    range: Range
+): Position {
+    let blockElement = getBlockElementAtNode(root, position.node);
+
+    if (blockElement) {
+        // Find the first <A> tag within current block which covers current selection
+        // If there are more than one nested, let's handle the first one only since that is not a common scenario.
+        let anchor = queryElements(
+            root,
+            'a[href]',
+            null /*forEachCallback*/,
+            QueryScope.OnSelection,
+            createRange(position)
+        ).filter((a: HTMLElement) => blockElement.contains(a))[0];
+
+        // If this is about to insert node to an empty A tag, clear the A tag and reset position
+        if (anchor && isNodeEmpty(anchor)) {
+            position = new Position(anchor, PositionType.Before);
+            safeRemove(anchor);
+            anchor = null;
+        }
+
+        // If this is about to insert nodes which contains A tag into another A tag, need to break current A tag
+        // otherwise we will have nested A tags which is a wrong HTML structure
+        if (anchor && safeInstanceOf(nodeToInsert, 'HTMLElement')) {
```

Comment:
My bad, after change this to use safeInstanceOf(), it has a regression become when paste, the nodeToInsert is actually a "DocumentFragment". Please change it back.